// cssom/lib/CSSConditionRule.js
import { __exports as CSSConditionRule } from '../../../node_modules/_virtual/CSSConditionRule.js';
export { __exports as default } from '../../../node_modules/_virtual/CSSConditionRule.js';
import './CSSRule.js';
import './CSSGroupingRule.js';
import { __exports as CSSRule } from '../../../node_modules/_virtual/CSSRule.js';
import { __exports as CSSGroupingRule } from '../../../node_modules/_virtual/CSSGroupingRule.js';

//.CommonJS
var CSSOM = {
  CSSRule: CSSRule.CSSRule,
  CSSGroupingRule: CSSGroupingRule.CSSGroupingRule
}; ///CommonJS

/**
 * @constructor
 * @see https://www.w3.org/TR/css-conditional-3/#the-cssconditionrule-interface
 */

CSSOM.CSSConditionRule = function CSSConditionRule() {
  CSSOM.CSSGroupingRule.call(this);
  this.cssRules = [];
};

CSSOM.CSSConditionRule.prototype = new CSSOM.CSSGroupingRule();
CSSOM.CSSConditionRule.prototype.constructor = CSSOM.CSSConditionRule;
CSSOM.CSSConditionRule.prototype.conditionText = '';
CSSOM.CSSConditionRule.prototype.cssText = ''; //.CommonJS

CSSConditionRule.CSSConditionRule = CSSOM.CSSConditionRule; ///CommonJS
