// cssom/lib/CSSKeyframeRule.js
import { __exports as CSSKeyframeRule } from '../../../node_modules/_virtual/CSSKeyframeRule.js';
export { __exports as default } from '../../../node_modules/_virtual/CSSKeyframeRule.js';
import './CSSRule.js';
import './CSSStyleDeclaration.js';
import { __exports as CSSRule } from '../../../node_modules/_virtual/CSSRule.js';
import { __exports as CSSStyleDeclaration } from '../../../node_modules/_virtual/CSSStyleDeclaration.js';

//.CommonJS
var CSSOM = {
  CSSRule: CSSRule.CSSRule,
  CSSStyleDeclaration: CSSStyleDeclaration.CSSStyleDeclaration
}; ///CommonJS

/**
 * @constructor
 * @see http://www.w3.org/TR/css3-animations/#DOM-CSSKeyframeRule
 */

CSSOM.CSSKeyframeRule = function CSSKeyframeRule() {
  CSSOM.CSSRule.call(this);
  this.keyText = '';
  this.style = new CSSOM.CSSStyleDeclaration();
  this.style.parentRule = this;
};

CSSOM.CSSKeyframeRule.prototype = new CSSOM.CSSRule();
CSSOM.CSSKeyframeRule.prototype.constructor = CSSOM.CSSKeyframeRule;
CSSOM.CSSKeyframeRule.prototype.type = 8; //FIXME
//CSSOM.CSSKeyframeRule.prototype.insertRule = CSSStyleSheet.prototype.insertRule;
//CSSOM.CSSKeyframeRule.prototype.deleteRule = CSSStyleSheet.prototype.deleteRule;
// http://www.opensource.apple.com/source/WebCore/WebCore-955.66.1/css/WebKitCSSKeyframeRule.cpp

Object.defineProperty(CSSOM.CSSKeyframeRule.prototype, "cssText", {
  get: function () {
    return this.keyText + " {" + this.style.cssText + "} ";
  }
}); //.CommonJS

CSSKeyframeRule.CSSKeyframeRule = CSSOM.CSSKeyframeRule; ///CommonJS
